ntb_hw_switchtec.c

```diff
@@ -1581,6 +1638,104 @@ static int switchtec_ntb_reinit_peer(struct switchtec_ntb *sndev)
 	return rc;
 }
 
+static ssize_t add_requester_id_store(struct device *dev,
+				      struct device_attribute *attr,
+				      const char *buf, size_t count)
+{
+	struct ntb_dev *ntb = container_of(dev, struct ntb_dev, dev);
+	struct switchtec_ntb *sndev = ntb_sndev(ntb);
+	int req_id;
+	int bus, device, func;
+	int rc;
+
+	if (sscanf(buf, "%x:%x.%x", &bus, &device, &func) != 3)
+		return -EINVAL;
+
+	req_id = PCI_DEVID(bus, PCI_DEVFN(device, func));
+	rc = add_req_id(sndev, sndev->mmio_self_ctrl, req_id);
+	if (rc)
+		return rc;
+
+	if (crosslink_is_enabled(sndev)) {
+		rc = crosslink_setup_req_ids(sndev, sndev->mmio_peer_ctrl);
+		if (rc)
+			return rc;
+	}
+
+	return count;
+}
+static DEVICE_ATTR_WO(add_requester_id);
+
+static ssize_t del_requester_id_store(struct device *dev,
+				      struct device_attribute *attr,
+				      const char *buf, size_t count)
+{
+	struct ntb_dev *ntb = container_of(dev, struct ntb_dev, dev);
+	struct switchtec_ntb *sndev = ntb_sndev(ntb);
+	int req_id;
+	int bus, device, func;
+	int rc;
+
+	if (sscanf(buf, "%x:%x.%x", &bus, &device, &func) != 3)
+		return -EINVAL;
+
+	req_id = PCI_DEVID(bus, PCI_DEVFN(device, func));
+	rc = del_req_id(sndev, sndev->mmio_self_ctrl, req_id);
+	if (rc)
+		return rc;
+
+	if (crosslink_is_enabled(sndev)) {
+		rc = crosslink_setup_req_ids(sndev, sndev->mmio_peer_ctrl);
+		if (rc)
+			return rc;
+	}
+
+	return count;
+}
+static DEVICE_ATTR_WO(del_requester_id);
+
+static ssize_t requester_ids_show(struct device *dev,
+				  struct device_attribute *attr, char *buf)
+{
+	struct ntb_dev *ntb = container_of(dev, struct ntb_dev, dev);
+	struct switchtec_ntb *sndev = ntb_sndev(ntb);
+	int i;
+	int table_size;
+	char req_id_str[32];
+	u32 req_id;
+	ssize_t n = 0;
+
+	table_size = ioread16(&sndev->mmio_self_ctrl->req_id_table_size);
+
+	for (i = 0; i < table_size; i++) {
+		req_id = ioread32(&sndev->mmio_self_ctrl->req_id_table[i]);
+
+		if (!(req_id & NTB_CTRL_REQ_ID_EN)) {
+			n += sprintf(req_id_str, "Requester ID #%d: N/A\n", i);
+		} else {
+			req_id >>= 16;
+			n += sprintf(req_id_str, "Requester ID #%d: %02x:%02x.%x\n", i,
```

Comment:
Generally we don't want our sysfs files to be so cluttered. The usual rule of thumb is each attribute should be one value.

However, that's not really possible here. 

But, I'd certainly not include "Requester ID" on ever line. Probably only just the enabled requester IDs, maybe the index in hardware in a table

"%d\t%02x:%02x.%x"

I certainly wouldn't print the unenabled slots. That's not useful information and it'd be a pain to parse.